### ISSUE TITLE
new check: UnnecessarySemicolonInTryWithResources

### ISSUE THREAD

**Issue by strkkk** [checkstyle/checkstyle#6774]
From #6752 
For last resource in resources declaration semicolon is unnecessary.
It should be detected and reported as violation.
If closing paren is not on same line, no violation in default configuration.
```
$ cat Test.java
import java.io.*;

public class TestClass {
  void method(){
       try( OutputStream s = new PipedOutputStream();)  // violation
       {}
       try(
           OutputStream s1 = new PipedOutputStream();
           OutputStream s2 = new PipedOutputStream();)  // violation
       {}
       try(
           OutputStream s1 = new PipedOutputStream();
           OutputStream s2 = new PipedOutputStream(); // no violation, closing paren on next line
       ){}
   }
}; 

$ cat conf.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
       <module name="UnnecessarySemicolonInTryWithResources" />
    </module>
</module>

$ java -jar checkstyle-8.20-all.jar -c conf.xml Test.java
Starting audit...
here is violations at lines commented with // violation
```
Additional boolean property `allowWhenNoBraceAfterSemicolon` (default values is true) to make an exception for lines, where closing parentheses is on the next line:
```
$ cat Test.java
import java.io.*;

public class TestClass {
  void method(){
       try( 
            OutputStream s = new PipedOutputStream();)  // violation
       {}
       try(
           OutputStream s1 = new PipedOutputStream();
           OutputStream s2 = new PipedOutputStream();)  // violation
       {}
       try(
           OutputStream s = new PipedOutputStream(); // violation
           ) {}
       try(
           OutputStream s1 = new PipedOutputStream();
           OutputStream s2 = new PipedOutputStream(); // violation 
       ){}
   }
}; 

$ cat conf.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
       <module name="UnnecessarySemicolonInTryWithResources">
          <property name="allowWhenNoBraceAfterSemicolon" value="false"/>
       </module>
    </module>
</module>

$ java -jar checkstyle-8.20-all.jar -c conf.xml Test.java
Starting audit...
here is violations at lines commented with // violation
```

**Comment by romani** [comment]
with such examples, I would rather enforce extra ";" to minimize diff on new resource addition to TRY.
If you want to have such Check you need to be clear that it is valuable when try is singe-line (or have a property foe this).
To not conflict with future Check that will demand extra ";" if ")" of TRY is on separate line.

**Comment by strkkk** [comment]
@romani I didn't understand your point.
Did you mean to add a property to make an exception for cases where closing ')' is on the next line?

**Comment by romani** [comment]
yes, lets reconcile conflict of interests by special property `allowWhenNoBraceAfterSemicolon=(true|false)` with default value to be `true`.

In your personal style you will use `allowWhenNoBraceAfterSemicolon=false`. 

for `allowWhenNoBraceAfterSemicolon=true`:
```
       try(
           OutputStream s = new PipedOutputStream(); // no violation
       ){}
       try(
           OutputStream s1 = new PipedOutputStream();
           OutputStream s2 = new PipedOutputStream(); // no violation
       ){}
       try(
           OutputStream s = new PipedOutputStream();) // violation
       {}
       try(
           OutputStream s1 = new PipedOutputStream();
           OutputStream s2 = new PipedOutputStream();) // no violation
       {}
       try(OutputStream s = new PipedOutputStream();) { // violation
       }
       try(Stream s1 = new Stream();Stream s2 = new Stream();){ // violation
       }
```
----

you can read more on why trailing comma is good at https://checkstyle.org/config_coding.html#ArrayTrailingComma , the same is true for `;` (minimization of code diff).

----

jdk spec - https://docs.oracle.com/javase/specs/jls/se8/html/jls-14.html#jls-14.20.3

**Comment by strkkk** [comment]
@romani Ok, sounds reasonable, thank you.
I updated first post with description and examples with `allowWhenNoBraceAfterSemicolon` property.
Please check.

**Comment by romani** [comment]
I am ok with this proposal.
@rnveach and @pbludov , please  your thoughts on this

**Comment by rnveach** [comment]
I am ok with proposal.

**Comment by romani** [comment]
@pbludov , even we already have two approval ,please review.

**Comment by romani** [comment]
@pbludov , even we already have two approval ,please review.

**Comment by pbludov** [comment]
I am ok with this proposal.

**Comment by rnveach** [comment]
Module was merged
